docker-compose-rule-core/src/main/java/com/palantir/docker/compose/execution/DockerCommandLocations.java

```diff
@@ -18,26 +18,40 @@
 import static java.util.Arrays.asList;
 
 import java.io.File;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import java.util.function.Predicate;
+import java.util.stream.Stream;
+import org.apache.commons.lang3.StringUtils;
 
 public class DockerCommandLocations {
     private static final Predicate<String> IS_NOT_NULL = path -> path != null;
     private static final Predicate<String> FILE_EXISTS = path -> new File(path).exists();
 
     private final List<String> possiblePaths;
+    private final String exeName;
+    private final boolean lookInPath;
 
-    public DockerCommandLocations(String... possiblePaths) {
+    public DockerCommandLocations(String exeName, boolean lookInPath, String... possiblePaths) {
```

Comment:
I really like this idea, it should reduce the amount of configuration necessary for people to use the library. However, this constructor now seems to be getting quite complicated and I don't think it's possible to write a test for finding the executable in the path with how this is written.

This seems like a really good candidate for using Immutables (https://immutables.github.io/), which we use elsewhere in the codebase - see DockerComposeExecArgument for a simple example. Using immutables you could do something like this:

```
@Value.Immutable
public abstract class DockerCommandLocations {

  protected String executableName();
  protected List<String> additionalSearchLocations();

  @Value.Default
   protected List<String> path() {
      // read path environment variable
   }

  @Value.Derived
  public Optional<String> preferredLocation() { ... }

  public static Builder builder();

}
```

Then, using the builder, it is possible to write unit tests by setting the path @Value.Default value so we can ensure this continues working in the future.